gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpoint.java

```diff
@@ -210,41 +212,51 @@ private void saveEnrollment(RoutingContext routingContext) {
                         routingContext.session().put(ConstantKeys.MFA_CHALLENGE_COMPLETED_KEY, true);
                         redirectToAuthorize(routingContext);
                     });
-            return;
+            return true;
         }
+        return false;
+    }
+
 
+    private Optional<Map.Entry<io.gravitee.am.model.Factor, FactorProvider>> getValidFactor(RoutingContext routingContext, String factorId, Client client) {
         if (factorId == null) {
             logger.warn("No factor id in form - did you forget to include factor id value ?");
             routingContext.fail(400);
-            return;
+            return Optional.empty();
         }
 
         final Map<io.gravitee.am.model.Factor, FactorProvider> factors = getFactors(client);
         Optional<Map.Entry<io.gravitee.am.model.Factor, FactorProvider>> optFactor = factors.entrySet().stream().filter(factor -> factorId.equals(factor.getKey().getId())).findFirst();
         if (optFactor.isEmpty()) {
             logger.warn("Factor not found - did you send a valid factor id ?");
             routingContext.fail(400);
-            return;
+            return Optional.empty();
         }
 
         if (routingContext.user() == null) {
             logger.warn("User must be authenticated to enroll MFA challenge.");
             routingContext.fail(401);
-            return;
+            return Optional.empty();
         }
 
         final var endUser = ((io.gravitee.am.gateway.handler.common.vertx.web.auth.user.User) routingContext.user().getDelegate()).getUser();
         if (userAlreadyHasFactor(endUser, client, routingContext.session())){
             logger.warn("User already has active factor, enrollment of factor '{}' rejected", factorId);
             routingContext.fail(new InvalidRequestException("factor already enrolled"));
-            return;
+            return Optional.empty();
         }
+        return optFactor;
+    }
 
-        // manage enrolled factors
-        FactorProvider provider = optFactor.get().getValue();
-        if (provider.checkSecurityFactor(getSecurityFactor(params, optFactor.get().getKey()))) {
+    private void manageEnrolledFactors(RoutingContext routingContext, Map.Entry<io.gravitee.am.model.Factor, FactorProvider> optFactor, MultiMap params) {
```

Comment:
**todo**: looks like you are not using the FactorProvider (optFactor.value), can yopu please only provide the Factor object and rename the param factor?

Reply:
it is used:
        FactorProvider provider = optFactor.getValue();
        if (provider.checkSecurityFactor(getSecurityFactor(params, optFactor.getKey()))) {